### PR TITLE
Preview Release ConnectedMachine 2024-03-31

### PR DESCRIPTION
**Pull request by yaotongms**
## Description

This release is targeting version 2024-03-31-preview of ConnectedMachine.

Changelog: 
Add commands that operate on ESU license:
- Get-AzConnectedLicense: get a list or a specific ESU license
- New-AzConnectedLicense: create a new ESU license
- Remove-AzConnectedLicense: remove an ESU license
- Set-AzConnectedLicense: update an ESU license

## Mandatory Checklist

- Please choose the target release of Azure PowerShell. (⚠️**Target release** is a different concept from **API readiness**. Please click below links for details.)
  - [ ] [General release](../blob/main/CONTRIBUTING.md#target-release-types)
  - [x] [Public preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Private preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Engineering build](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] No need for a release

- [ ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For SDK-based development mode, update `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`.
        * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. 
    * For autorest-based development mode, include the changelog in the PR description.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: src/ConnectedMachine/ConnectedMachine.Autorest/README.md [hunk 1]
```diff
@@ -85,102 +85,104 @@ directive:
...
  
-  - from: swagger-document
-    where: $.definitions.Machine.properties.properties
-    transform: >-
-      return {
-          "x-ms-client-flatten": true,
-          "$ref": "#/definitions/MachineProperties",
-          "description": "Hybrid Compute Machine properties"
-        }
+  # - from: swagger-document
```
Comment (YanaXu) [inline src/ConnectedMachine/ConnectedMachine.Autorest/README.md:88]: If it's not used anymore, please remove it.

**Comment by YanaXu** [comment]
Please fix CI errors.

File: src/ConnectedMachine/ConnectedMachine.Autorest/README.md (hunk 1 above)
Comment (yaotongms) [reply src/ConnectedMachine/ConnectedMachine.Autorest/README.md]: fixed

File: src/ConnectedMachine/ConnectedMachine.Autorest/docs/New-AzConnectedLicense.md [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+ [-SubscriptionId <String>] [-LicenseDetailEdition <String>] [-LicenseDetailProcessor <Int32>]
+ [-LicenseDetailState <String>] [-LicenseDetailTarget <String>] [-LicenseDetailType <String>]
+ [-LicenseDetailVolumeLicenseDetail <IVolumeLicenseDetails[]>] [-LicenseType <String>] [-Tag <Hashtable>]
+ [-TenantId <String>] [-DefaultProfile <PSObject>] [-AsJob] [-NoWait] [-Confirm] [-WhatIf]
+ [<CommonParameters>]
+```
+
+### Create
+```
+New-AzConnectedLicense -Name <String> -ResourceGroupName <String> -Parameter <ILicense>
```
Comment (YanaXu) [inline src/ConnectedMachine/ConnectedMachine.Autorest/docs/New-AzConnectedLicense.md:27]: Add a model cmdlet for "Parameter".

File: src/ConnectedMachine/ConnectedMachine.Autorest/test/Get-AzConnectedExtensionMetadata.Tests.ps1 [hunk 3]
```diff
@@ -12,13 +12,13 @@ while(-not $mockingPath) {
 . ($mockingPath | Select-Object -First 1).FullName
 
 Describe 'Get-AzConnectedExtensionMetadata' {
-    It 'List' {
-        $all = @(Get-AzConnectedExtensionMetadata -ExtensionType 'CustomScriptExtension' -Publisher 'Microsoft.Compute' -Location 'eastus2euap')
+    It 'List' -skip{
```
Comment (YanaXu) [inline src/ConnectedMachine/ConnectedMachine.Autorest/test/Get-AzConnectedExtensionMetadata.Tests.ps1:15]: why to skip this  test?

File: src/ConnectedMachine/ConnectedMachine.Autorest/test/Get-AzConnectedLicense.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  $currentPath = $PSScriptRoot
+  while(-not $mockingPath) {
+      $mockingPath = Get-ChildItem -Path $currentPath -Recurse -Include 'HttpPipelineMocking.ps1' -File
+      $currentPath = Split-Path -Path $currentPath -Parent
+  }
+  . ($mockingPath | Select-Object -First 1).FullName
+}
+
+Describe 'Get-AzConnectedLicense' {
+    It 'List' -skip{
```
Comment (YanaXu) [inline src/ConnectedMachine/ConnectedMachine.Autorest/test/Get-AzConnectedLicense.Tests.ps1:18]: All tests are skipped. It should not be.

File: src/ConnectedMachine/ConnectedMachine.Autorest/README.md [hunk 5]
```diff
@@ -367,7 +386,11 @@ directive:
         description: Gets the list of ResourceGroupName's available for this subscription.
         script: Get-AzResourceGroup | Select-Object -ExpandProperty ResourceGroupName
  
-  # These APIs are used by the agent so they do not need to be in the cmdlets.
+  # These APIs are used by the agent so they do not need to be in the cmdlets
   - remove-operation: Machines_CreateOrUpdate
   - remove-operation: MachineRunCommands_Update
+
+  # Create model cmdlet for complex object
+  - model-cmdlet:
```
Comment (YanaXu) [inline src/ConnectedMachine/ConnectedMachine.Autorest/README.md:394]: The right way for "model-cmdlet" is 
```yaml
- model-cmdlet:
    - model-name: ExtensionValue
      cmdlet-name: New-AzConnectedExtensionValue
```

File: src/ConnectedMachine/ConnectedMachine.Autorest/docs/New-AzConnectedLicense.md [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+New-AzConnectedLicense -InputObject <IConnectedMachineIdentity> -Parameter <ILicense>
+ [-DefaultProfile <PSObject>] [-AsJob] [-NoWait] [-Confirm] [-WhatIf] [<CommonParameters>]
+```
+
+### CreateViaIdentityExpanded
+```
+New-AzConnectedLicense -InputObject <IConnectedMachineIdentity> -Location <String>
+ [-LicenseDetailEdition <String>] [-LicenseDetailProcessor <Int32>] [-LicenseDetailState <String>]
+ [-LicenseDetailTarget <String>] [-LicenseDetailType <String>]
+ [-LicenseDetailVolumeLicenseDetail <IVolumeLicenseDetails[]>] [-LicenseType <String>] [-Tag <Hashtable>]
```
[the remark this answers is not in the document]
Comment (yaotongms) [reply src/ConnectedMachine/ConnectedMachine.Autorest/docs/New-AzConnectedLicense.md:43]: added for license detail

File: src/ConnectedMachine/ConnectedMachine.Autorest/docs/New-AzConnectedLicense.md (hunk 2 above)
Comment (yaotongms) [reply src/ConnectedMachine/ConnectedMachine.Autorest/docs/New-AzConnectedLicense.md:27]: added for license detail

File: src/ConnectedMachine/ConnectedMachine.Autorest/examples/Get-AzConnectedNetworkSecurityPerimeterConfiguration.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
+### Example 1: {{ Add title here }}
```
Comment (YanaXu) [inline src/ConnectedMachine/ConnectedMachine.Autorest/examples/Get-AzConnectedNetworkSecurityPerimeterConfiguration.md:1]: Please update this md file.
Fill in all placeholders.

File: src/ConnectedMachine/ConnectedMachine.Autorest/examples/Invoke-AzConnectedReconcileNetworkSecurityPerimeterConfiguration.md [hunk 8]
```diff
@@ -0,0 +1,22 @@
+### Example 1: {{ Add title here }}
```
Comment (YanaXu) [inline src/ConnectedMachine/ConnectedMachine.Autorest/examples/Invoke-AzConnectedReconcileNetworkSecurityPerimeterConfiguration.md:1]: Fill in all placeholders.

File: src/ConnectedMachine/ConnectedMachine.Autorest/test/Get-AzConnectedMachine.Tests.ps1 [hunk 9]
```diff
@@ -1,65 +1,27 @@
...
-    It 'Get all connected machines in a subscription' {
-        $machines = Get-AzConnectedMachine
-        $machines.Count | Should -BeGreaterThan 1
-    }
-
-    It 'Get all connected machines in a resource group' {
-        $machines = Get-AzConnectedMachine -ResourceGroupName $env.ResourceGroupName
-        $machines.Count | Should -Be 1
+    It 'List'  {
+        $machine = Get-AzConnectedMachine
```
Comment (YanaXu) [inline src/ConnectedMachine/ConnectedMachine.Autorest/test/Get-AzConnectedMachine.Tests.ps1:19]: Please record this test.

File: src/ConnectedMachine/ConnectedMachine.Autorest/test/Invoke-AzConnectedReconcileNetworkSecurityPerimeterConfiguration.Tests.ps1 [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  $TestRecordingFile = Join-Path $PSScriptRoot 'Invoke-AzConnectedReconcileNetworkSecurityPerimeterConfiguration.Recording.json'
+  $currentPath = $PSScriptRoot
+  while(-not $mockingPath) {
+      $mockingPath = Get-ChildItem -Path $currentPath -Recurse -Include 'HttpPipelineMocking.ps1' -File
+      $currentPath = Split-Path -Path $currentPath -Parent
+  }
+  . ($mockingPath | Select-Object -First 1).FullName
+}
+
+Describe 'Invoke-AzConnectedReconcileNetworkSecurityPerimeterConfiguration' {
```
Comment (YanaXu) [inline src/ConnectedMachine/ConnectedMachine.Autorest/test/Invoke-AzConnectedReconcileNetworkSecurityPerimeterConfiguration.Tests.ps1:17]: Please record this again.

File: src/ConnectedMachine/ConnectedMachine.Autorest/test/Set-AzConnectedMachineExtension.Tests.ps1 [hunk 11]
```diff
@@ -1,18 +1,47 @@
...
-$currentPath = $PSScriptRoot
-while(-not $mockingPath) {
-    $mockingPath = Get-ChildItem -Path $currentPath -Recurse -Include 'HttpPipelineMocking.ps1' -File
-    $currentPath = Split-Path -Path $currentPath -Parent
-}
-. ($mockingPath | Select-Object -First 1).FullName
-
-Import-Module "$PSScriptRoot/helper.psm1" -Force
 
 Describe 'Set-AzConnectedMachineExtension' {
```
Comment (YanaXu) [inline src/ConnectedMachine/ConnectedMachine.Autorest/test/Set-AzConnectedMachineExtension.Tests.ps1:17]: Please  record this again.

File: src/ConnectedMachine/ConnectedMachine.Autorest/test/Update-AzConnectedMachineExtension.Tests.ps1 [hunk 12]
```diff
@@ -1,107 +1,53 @@
...
     }
 
-    It 'Update parameter set' {
-        $currentExtension = Get-AzConnectedMachineExtension -ResourceGroupName $env.ResourceGroupName -MachineName $machineName -Name custom1
-        $newCommand = "powershell.exe echo hi"
-        $currentExtension.Setting["commandToExecute"] = $newCommand
+    It 'Update' {
+        $target = @{"Microsoft.Compute.CustomScriptExtension" = @{"targetVersion"="1.10.12"}}
+        $all = (Update-AzConnectedExtension -ResourceGroupName $env.ResourceGroupName -MachineName $env.MachineName -ExtensionTarget $target)
+        $all | Should -Not -BeNullOrEmpty
```
Comment (YanaXu) [inline src/ConnectedMachine/ConnectedMachine.Autorest/test/Update-AzConnectedMachineExtension.Tests.ps1:33]: Double check this test. It fails.

File: src/ConnectedMachine/ConnectedMachine.Autorest/examples/Get-AzConnectedNetworkSecurityPerimeterConfiguration.md (hunk 7 above)
Comment (yaotongms) [reply src/ConnectedMachine/ConnectedMachine.Autorest/examples/Get-AzConnectedNetworkSecurityPerimeterConfiguration.md]: removed from this release

File: src/ConnectedMachine/ConnectedMachine.Autorest/examples/Invoke-AzConnectedReconcileNetworkSecurityPerimeterConfiguration.md (hunk 8 above)
Comment (yaotongms) [reply src/ConnectedMachine/ConnectedMachine.Autorest/examples/Invoke-AzConnectedReconcileNetworkSecurityPerimeterConfiguration.md]: removed in this release

File: src/ConnectedMachine/ConnectedMachine.Autorest/test/Invoke-AzConnectedReconcileNetworkSecurityPerimeterConfiguration.Tests.ps1 (hunk 10 above)
Comment (yaotongms) [reply src/ConnectedMachine/ConnectedMachine.Autorest/test/Invoke-AzConnectedReconcileNetworkSecurityPerimeterConfiguration.Tests.ps1]: removed in this release

**Review by YanaXu — APPROVED** [review]

**Pull request merged**